### ISSUE TITLE
Allow creating a custom resource that inherits from a core resource

### ISSUE THREAD

**Issue by KrisShannon** [inspec/inspec#4328]
## Detailed Description
Sometimes customisations of core resources are necessary and don't make sense to upstream.

Unfortunately just inheriting from a core resource class in a profile library file doesn't work as the resource is registered into the wrong registry.

## Context
Two instances where I would have liked to use inheritance:

* Currently `grub_conf` doesn't provide a way to force the use of the legacy parser (see #4323)  This may be eventually fixed but until then an easy workaround would be a simple inherited `grub_legacy_conf` that would only need to override `initialize`
* We have a completely custom build of nginx that has some other custom kinds of grouping.  Inheriting from NginxConf would be particularly useful here as the changes would not be appropriate for upstreaming

## Possible Implementation
Maybe allow a second argument to `Inspec.resource` with a requested superclass:

```ruby
   def self.resource(version, superclass = nil)
     validate_resource_dsl_version!(version)
     if superclass.nil?
       Inspec::Plugins::Resource
     else
       Class.new(superclass) do
          def __resource_registry
            Inspec::Resource.registry
          end
       end
     end
   end
```

(I haven't checked if that gets the namespacing correct - it's just off the top of my head)

Aha! Link: https://chef.aha.io/features/SH-2700

**Comment by KrisShannon** [comment]
Just for reference,  a hack that I am currently using for the `grub_conf` is

```ruby
PROFILE_RESOURCE_REGISTRY = Inspec.resource(1).__resource_registry

class GrubLegacyConfig < GrubConfig
  def self.__resource_registry
   PROFILE_RESOURCE_REGISTRY
  end
  name 'grub_legacy_conf'
  def initialize(path = nil, kernel = nil)
    @conf_path = path || '/boot/grub/grub.conf'
    @version = 'legacy'
    @content = read_file(@conf_path)
    @kernel = kernel || 'default'
  end
end
```

**Comment by KrisShannon** [comment]
Previous closed tickets I found about inheritance:

* #1300 - custom resource subclass of existing inpec resource no longer works
* #1613 - Resource class inheritance error

**Comment by mNantern** [comment]
@KrisShannon do you know if your workaround should work with resources coming from a profile (like [inspec-aws](https://github.com/inspec/inspec-aws)) ?

When I try it only override the core resource (which has the same name unfortunately). Do you know a way to override a profile Resource ?

**Comment by KrisShannon** [comment]
@mNantern There has been a massive overhaul with how resources are loaded to enable lazy loading and my workaround no longer works at all.

Until inheritance becomes officially supported we’ve ended up copying the whole resource and patching instead.

**Comment by mNantern** [comment]
That's sad but thank you !

**Comment by pgmillon** [comment]
Hi, any updates on this ? It's the first issue listed when searching for aws resource.

**Comment by pgmillon** [comment]
A workaround we found and works well is to define a module and extend the Core classes:

test/integration/base_profile/libraries/aws_mixins.rb
```ruby

module AwsIamRoleMixin
  def inline_policies_names
    @aws.iam_client.list_role_policies(role_name: @role_name).policy_names
  end
  def inline_policies
    inline_policies_names.map do |policy_name|
      JSON.parse(
          CGI::unescape(
              @aws.iam_client.get_role_policy(role_name: @role_name, policy_name: policy_name).policy_document
          )
      )
    end
  end
end

class AwsMixins < Inspec.resource(1)
  name 'aws_mixins'

  def AwsIamRoleMixin
    AwsIamRoleMixin
  end

end
```

And we can now read the inline policies without duplicating the whole resource:

```ruby
awsMixins = aws_mixins


control "iam-terraform-1.0" do
  impact 1.0
  title "Customer IAM role validation"

  describe aws_iam_role(role_name: ncp.customer_iam_role_name).extend awsMixins.AwsIamRoleMixin do
	its('inline_policies.count') { should eq 5 }
  end

end
```

There may be better ways, but at least it does the job for us.

**Comment by KrisShannon** [comment]
This workaround is ok,   but doesn't help for my initial use case because the `initialize` method has already been called before the `extend` can run.

**Comment by Vasu1105** [comment]
Fixed in #5816
